profiling/src/profiling.rs

```diff
@@ -476,41 +474,57 @@ impl TimeCollector {
+                            ProfilerMessage::Cancel => {
+                                // flush what we have before exiting
+                                last_wall_export = self.handle_timeout(&mut profiles, &last_wall_export);
+                                running = false;
+                            }
+                        },
+                        Err(_) => {
+                            /* Docs say:
+                             * > A message could not be received because the
+                             * > channel is empty and disconnected.
+                             * If this happens, let's just break and end.
+                             */
                             break;
                         }
                     }
                 },
-                recv(self.cancel_receiver) -> _ => {
-                    // flush what we have before exiting
-                    last_wall_export = self.handle_timeout(&mut profiles, &last_wall_export);
-                    running = false;
-                },
-                recv(wall_time_tick) -> _ => {
-                    match self.vm_interrupt_lock.lock() {
-                        Ok(interrupts) => {
-                            interrupts.iter().for_each(|obj| unsafe {
-                                (&*obj.engine_ptr).store(true, Ordering::SeqCst);
-                                (&*obj.interrupt_count_ptr).fetch_add(1, Ordering::SeqCst);
-                            });
-                        }
-                        Err(err) => {
-                            error!(
-                                "Stopping time related profiling: failed to acquire vm_interrupt lock: {}",
-                                err
-                            );
-                            break;
-                        },
+
+                recv(wall_time_tick) -> message => {
+                    if message.is_ok() {
+                        let vm_interrupts = self.vm_interrupts.lock().unwrap();
+
+                        vm_interrupts.iter().for_each(|obj| unsafe {
+                            (&*obj.engine_ptr).store(true, Ordering::SeqCst);
+                            (&*obj.interrupt_count_ptr).fetch_add(1, Ordering::SeqCst);
+                        });
                     }
                 },
-                recv(upload_tick) -> _ => {
-                    last_wall_export = self.handle_timeout(&mut profiles, &last_wall_export);
+
+                recv(upload_tick) -> message => {
+                    if message.is_ok() {
+                        last_wall_export = self.handle_timeout(&mut profiles, &last_wall_export);
+                    }
                 },
+
+                recv(self.fork_receiver) -> message => {
+                    if message.is_ok() {
+                        // First, wait for every thread to finish what they are currently doing.
+                        self.fork_barrier.wait();
```

Comment:
I'm confused as to why we would always sync on the fork barrier ? Should this not only activate when we are forking ?

Reply:
We do this whenever we receive a message on the `fork_receiver`, which will only be sent when we are forking.